tensorboard/webapp/runs/effects/runs_effects.ts

```diff
@@ -240,37 +129,35 @@ export class RunsEffects {
     );
   }
 
-  private maybeWaitForRunsAndGetRuns(experimentId: string): Observable<{
-    fromRemote: false;
-    experimentId: string;
-    runs: Run[];
-  }> {
-    return this.store.select(getRunsLoadState, {experimentId}).pipe(
-      filter((loadState) => loadState.state !== DataLoadState.LOADING),
-      take(1),
-      mergeMap((loadState) => {
-        if (loadState.state === DataLoadState.FAILED) {
-          return throwError(new Error('Pending request failed'));
-        }
-        return of(loadState);
-      }),
-      withLatestFrom(this.store.select(getRuns, {experimentId})),
-      map(([, runs]) => ({fromRemote: false, experimentId, runs}))
+  private fetchRunsForExperiments(
+    experimentIds: string[]
+  ): Observable<Record<string, Run[]>> {
+    return forkJoin(
+      experimentIds.map((eid) =>
+        this.runsDataSource.fetchRuns(eid).pipe(map((run) => [eid, run]))
+      )
+    ).pipe(
+      map(([...eidRunPairs]) => {
+        return Object.fromEntries(eidRunPairs as Array<[string, Run[]]>);
+      })
     );
   }
 
-  private fetchRunsForExperiment(experimentId: string): Observable<{
-    fromRemote: true;
-    experimentId: string;
-    runs: Run[];
-    metadata: HparamsAndMetadata;
-  }> {
+  private fetchRunsAndHparamsForExperiments(
+    experimentIds: string[]
+  ): Observable<RunsAndMetadata[]> {
+    if (!experimentIds.length) return of([]);
     return forkJoin([
-      this.runsDataSource.fetchRuns(experimentId),
-      this.runsDataSource.fetchHparamsMetadata(experimentId),
+      this.runsDataSource.fetchHparamsMetadata(experimentIds),
+      this.fetchRunsForExperiments(experimentIds),
     ]).pipe(
-      map(([runs, metadata]) => {
-        return {fromRemote: true, experimentId, runs, metadata};
+      map(([metadata, eidToRuns]) => {
+        return experimentIds.map((experimentId) => ({
+          fromRemote: true,
+          experimentId,
+          runs: eidToRuns[experimentId],
+          metadata,
```

Comment:
At this point I think it starts to just work sort of accidentally. Can we change how we store hparams "metadata" in the underlying data model, too?

metadata is incorrect here. It defines the hparam metadata for a group of experiments in the comparison (ExpA, ExpB, ExpC, etc...) but you are mapping it to each experiment individually.

It can get weird really quickly, I think. For example, if user:
1. Creates a comparison of ExpA, ExpB, ExpC.
2. Removes ExpA from the comparison.
3. Navigates back to list and then views ExpD

Would this be the hparams state?
```
{
  specs: { 
    ExpA: <Metadata for ExpA, ExpB, ExpD>,
    ExpB: <Metadata for ExpB, ExpD>
    ExpC: <Metadata for ExpC>
    ExpD: <Metadata for ExpB, ExpD>
  }
}
```

It seems to work despite this because you are constantly updating the metadata whenever user views a runs table. It all seems quite accidental and brittle.

Idealy I think you would track hparams metadata for the dashboard runs table differently.

Reply:
Because the metadata object does not change per iteration of the loop the state would look like this
```
{
  specs: {
    ExpA: <Metadata for ExpA, ExpB, ExpC>,
    ExpB: <Metadata for ExpB, ExpC>,
    ExpC: <Metadata for ExpB, ExpC>,
    ExpD: <Metadata for ExpD>,
  }
}
```